### PR TITLE
Adding tutorial about custom loss in Gluon

### PR DESCRIPTION
**Pull request by NRauschmayr**
Adding a tutorial about custom loss functions in Gluon.  https://github.com/mli/new-docs/projects/1#card-15544723

### CODE REVIEW

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
+
+# Custom Loss functions in Gluon
+
+Loss functions ared used to train a neural network and to compute the difference between output and target variable. Consequently, they are a key element to successfully train models. Depending on the model, we may choose one or another function. MXNet's Gluon API provides most commonly used functions. For instance:
+
+- regression: [L1Loss](https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.loss.L1Loss.html), [L2Loss](https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.loss.L2Loss.html) 
+- classification: [SigmoidBinaryCrossEntropyLoss](https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.loss.SigmoidBinaryCrossEntropyLoss.html), [SoftmaxBinaryCrossEntropyLoss](https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.loss.SoftmaxBinaryCrossEntropyLoss.html) 
+- embeddings: [HingeLoss](https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.loss.HingeLoss.html)
+
+However, we may sometimes want to solve problems that require customized loss functions and this tutorial shows how we can do that in Gluon. We will show it at the example of contrastive loss which is typically used in Siamese networks.
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: customized loss functions; this tutorial shows how seems more natural than using 'and' without commas

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+from mxnet import autograd, gluon, nd
+from mxnet.gluon.loss import Loss
+import random
+```
+
+### What is contrastive Loss
+
+[Contrastive loss](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) is a distance-based loss function. During training pairs of images are fed into a model. If the images are similar the loss function will return 0, otherwise 1. 
+
+<img src="contrastive_loss.jpeg" width="400">
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: does this need to be legal xml, i.e.  /> instead of > ?

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import matplotlib.pyplot as plt
+import mxnet as mx
+from mxnet import autograd, gluon, nd
+from mxnet.gluon.loss import Loss
+import random
+```
+
+### What is contrastive Loss
+
+[Contrastive loss](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) is a distance-based loss function. During training pairs of images are fed into a model. If the images are similar the loss function will return 0, otherwise 1.
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: During training, (comma) pairs of

Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: If the images are similar, the loss function will return 0, otherwise it will return 1.
(adding commas and expanding the contraction)

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+<img src="contrastive_loss.jpeg" width="400">
+
+*Y* is a binary label indicating similarity between training images. Contrastive loss uses the Euclidean distance *D* between images. The contrastive loss is the sum of 2 terms: 
+ - the first one indicates the loss for a pair of similar points
+ - the second one indicates the loss for a pair of dissimilar points
+
+The loss function uses a margin *m* which is has the effect that dissimlar pairs only contribute if their loss is within a certain margin. 
+
+In order to implement such a customized loss function in Gluon, we only need to define a new class that is inhertiting from the Loss base class. We then define the contrastive loss in the ```hybrid_forward```. This function takes the images ```image1```, ```image2``` and the label which defines whether  ```image1``` and ```image2``` are similar (=0) or  dissimilar (=1). The input F is either an ```mxnet.ndarry``` or an ```mxnet.symbol``` if we hybridize the network. Gluon's Loss base class is in fact a HybridBlock. This means we can either run fully imperatively or symbolically. When we hybridize our custom loss function, we can get performance speedups.
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: inheriting

Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: is it ```hybrid_forward``` or `hybrid_forward` (one tilde or three?) same elsewhere?

Comment (vishaalkapoor) [reply Rsite/source/tutorial/CustomLossGluon.md]: one will be inline, three is on it's own line, correct?

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        loss = (1 - label) * distances_squared +  label * F.square(d)
+        loss = 0.5*loss
+        
+        return loss
+    
+loss = ContrastiveLoss(margin=6.0)
+```
+
+### Define the Siamese network
+A Siamese network consists of 2 identical networks, that share the same weights. They are trained on pair of images and each network processes one image. The label defines whether the pair of images is similar or not. The Siamese network learns to differentiate between two input images.
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: Would you be able to add a link to Siamese networks - a source for more information?

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+            image1_index, image1_tuple = random.choice(items_with_index)
+
+        image0 = super().__getitem__(image0_index)
+        image1 = super().__getitem__(image1_index)
+
+        return image0[0], image1[0], mx.nd.array(mx.nd.array([int(image1_tuple[1] != image0_tuple[1])]))
+
+    def __len__(self):
+        return super().__len__()
+
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: unnecessary blank space

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        return image0[0], image1[0], mx.nd.array(mx.nd.array([int(image1_tuple[1] != image0_tuple[1])]))
+
+    def __len__(self):
+        return super().__len__()
+    
+
+```
+
+We train the network on a subset of the data, the  *Tifinagh* alphabet. Once the model is trained we test it on the *Inuktitut* alphabet.
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: reference to inuktitut alphabet

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+fig, (ax0, ax1) = plt.subplots(ncols=2, figsize=(10, 5)) 
+ax0.imshow(img1.asnumpy()[:,:,0], cmap='gray')
+ax0.axis('off')
+ax1.imshow(img2.asnumpy()[:,:,0], cmap='gray')
+ax1.axis("off")
+plt.show()
+
+```
+
+    Same: False
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: Is Same: False an artifact? Should it be there?

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    ax0.imshow(img1.asnumpy()[0,0,:,:], cmap='gray')
+    ax0.axis('off')
+    ax1.imshow(img2.asnumpy()[0,0,:,:], cmap='gray')
+    ax1.axis("off")
+    plt.show()
+
+    break
+```
+
+    Euclidean Distance: 2.494767 Test label 1.0
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: Same with this one - this looks like the output of the python code. I think the output should be generated during compilation. I've been checking my code with this command (for nn.md)
`python build/md2ipynb.py guide/module/gluon/nn.md no.ipynb && jupyter notebook nn.ipynb`    
from the new-docs/python directory and making sure the output is as desired.

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+####  Intermediate loss values:
+In our example, we computed the square root of squared distances between 2 images: ```F.sqrt(distances_squared)```. If images are very similar we take the sqare root of a value close to 0, which can lead to *NaN* values. Adding a small epsilon to ```distances_squared``` avoids this problem.
+
+#### Shape of intermediate loss vectors:
+Computations like ```F.sum``` collapse dimensions and this could have side effects. In most cases having the wrong shape will lead to an error, as soon as we compare data with labels. But in some cases, we may be able to normally run the training, but it does not converge. For instance, if we don't set ```keepdims=True``` in our customized loss function, the example will run fine but not converge. 
+
+#### Differentiable
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: Since this is long, a summary paragraph or conclusion seems like it would tie things up nicely

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+
+![png](CustomLossGluon2.png)
+
+
+### Common Pitfalls with custom loss functions
+
+When customizing loss functions, we may encounter certain pitfalls. If the loss is not decreasing as expected or if forward/backward pass is crashing, then one should check the following:
+
+#### Activation function in the last layer:
```
Comment (vishaalkapoor) [inline Rsite/source/tutorial/CustomLossGluon.md]: some of the section headers end in colons and some don't.
Also the capitalization of Common Pitfalls with custom loss functions doesn't match the capitalization of other headers

**Comment by vishaalkapoor** [comment]
LGTM with one comment - double-check that the python code output isn't included as this will be generated as part of the build.

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+Verify whether the last network layer uses the correct activation function: for instance in binary classification tasks we need to apply a sigmoid on the output data. If we use this activation in the last layer and define a loss function like Gluon's SigmoidBinaryCrossEntropy, we would basically apply sigmoid twice and the loss would not converge as expected. If we don't define any activation function, Gluon will per default apply a linear activation.
+
+####  Intermediate loss values
+In our example, we computed the square root of squared distances between 2 images: `F.sqrt(distances_squared)`. If images are very similar we take the sqare root of a value close to 0, which can lead to *NaN* values. Adding a small epsilon to `distances_squared` avoids this problem.
+
+#### Shape of intermediate loss vectors
+In most cases having the wrong tensor shape will lead to an error, as soon as we compare data with labels. But in some cases, we may be able to normally run the training, but it does not converge. For instance, if we don't set `keepdims=True` in our customized loss function, the shape of the tensor changes. The example still runs fine but does not converge.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply Rsite/source/tutorial/CustomLossGluon.md]: we may be able to normally run the training, but it does not converge -> we may be able to train as normal but the loss never converges.

File: Rsite/source/tutorial/CustomLossGluon.md [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+####  Intermediate loss values
+In our example, we computed the square root of squared distances between 2 images: `F.sqrt(distances_squared)`. If images are very similar we take the sqare root of a value close to 0, which can lead to *NaN* values. Adding a small epsilon to `distances_squared` avoids this problem.
+
+#### Shape of intermediate loss vectors
+In most cases having the wrong tensor shape will lead to an error, as soon as we compare data with labels. But in some cases, we may be able to normally run the training, but it does not converge. For instance, if we don't set `keepdims=True` in our customized loss function, the shape of the tensor changes. The example still runs fine but does not converge. 
+
+If you encounter a similar problem, then it is useful to check the tensor shape after each computation step in the loss function.
```
Comment (thomelane) [inline Rsite/source/tutorial/CustomLossGluon.md]: tensor -> array
